File: src/UnwindMC.Tests/SourceTests/SourceTester.cs
using System;
using System.IO;
using System.Linq;
using NUnit.Framework;

namespace UnwindMC.Tests.SourceTests
{
    class SourceTester
    {
        public static void TestDecompiler(string code, string expected)
        {
            code = Trim(code);

            Console.WriteLine("===================================== CODE =====================================");
            Console.WriteLine(code);
            Console.WriteLine("");

            var asm = Disassemble(code);

            Console.WriteLine("===================================== ASM ======================================");
            Console.WriteLine(asm);
            Console.WriteLine("");

            var cppCode = Decompiler.decompileFunction(AnalysisHelper.Analyze(asm), 0x0);

            Console.WriteLine("==================================== RESULT ====================================");
            Console.WriteLine(cppCode);
            Console.WriteLine("");

            Assert.That(cppCode, Is.EqualTo(Trim(expected)));
        }

        private static string Disassemble(string test)
        {
            using (var tempDir = VcTools.CreateTempDirectory())
            {
                var cppPath = Path.Combine(tempDir.Path, "test.cpp");
                File.WriteAllText(cppPath, test);
                var objPath = VcTools.Compile(cppPath);
                return VcTools.Disassemble(objPath);
            }
        }

        private static string Trim(string expected)
        {
            var lines = expected.Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries);
            int indent = lines[0].TakeWhile(c => c == ' ').Count();
            expected = string.Join(Environment.NewLine, lines.Select(l => l.Substring(indent)));
            return expected;
        }
    }
}

File: src/UnwindMC.Tests/SourceTests/SourceTester.fs
module rec SourceTester

open System
open System.IO
open System.Linq
open NUnit.Framework

let testDecompiler (code: string) (expected: string): unit =
    let code = trim code

    printfn "===================================== CODE ====================================="
    printfn "%s" code
    printfn ""

    let asm = disassemble code

    printfn "===================================== ASM ======================================"
    printfn "%s" asm
    printfn ""

    let cppCode = Decompiler.decompileFunction (AnalysisHelper.analyze asm) 0uL

    printfn "==================================== RESULT ===================================="
    printfn "%s" cppCode
    printfn ""

    Assert.That(cppCode, Is.EqualTo(trim(expected)))

    use tempDir = VcTools.createTempDirectory ()
    compile tempDir.Path cppCode |> ignore

let private disassemble (code: string): string =
    use tempDir = VcTools.createTempDirectory ()
    let objPath = compile tempDir.Path code
    VcTools.disassemble objPath

let private compile (dir: string) (code: string): string =
    let cppPath = Path.Combine(dir, "test.cpp")
    File.WriteAllText(cppPath, code)
    VcTools.compile cppPath

let private trim (expected: string): string =
    let lines = expected.Split([| '\r'; '\n' |], StringSplitOptions.RemoveEmptyEntries)
    let indent = lines.[0].TakeWhile(fun c -> c = ' ').Count()
    System.String.Join(Environment.NewLine, lines.Select(fun l -> l.Substring(indent)))
